lib/os/onoff.c

```diff
@@ -0,0 +1,503 @@
+	bool no_wait = false;
+	int rv = validate_args(srv, cli, &no_wait);
+
+	if (rv != 0) {
+		return rv;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+
+	if ((srv->flags & ONOFF_SERVICE_HAS_ERROR) != 0) {
+		rv = -EIO;
+		goto out;
+	}
+
+	u32_t state = srv->flags & SERVICE_STATE_MASK;
+
+	switch (state) {
+	case SERVICE_STATE_ON:
+		/* Stay on if release leaves a client. */
+		if (srv->refs > 1U) {
+			notify = true;
+			rv = 1;
+			break;
+		}
+
+		/* Reject if in isr but stop not isr-callable */
+		if (k_is_in_isr()
+		    && ((srv->flags & ONOFF_SERVICE_ISR_CALLABLE_STOP)
+			== 0)) {
+			rv = -EWOULDBLOCK;
+			break;
+		}
+
+		/* Reject if in no-wait condition and stop could
+		 * suspend.
+		 */
+		if (no_wait
+		    && ((srv->flags & ONOFF_SERVICE_SUSPENDABLE_STOP)
+			!= 0)) {
+			rv = -EWOULDBLOCK;
+			break;
+		}
+
+		stop = true;
+
+		set_service_state(srv, SERVICE_STATE_TO_OFF);
+		srv->releaser = cli;
+		rv = 2;
+
+		break;
+	case SERVICE_STATE_TO_ON:
+		rv = -EBUSY;
+		break;
+	case SERVICE_STATE_OFF:
+	case SERVICE_STATE_TO_OFF:
+		rv = -EALREADY;
+		break;
+	default:
+		rv = -EINVAL;
+	}
+
+out:
+	if (notify) {
+		srv->refs -= 1U;
+	}
+
+	k_spin_unlock(&srv->lock, key);
+
+	if (stop) {
+		__ASSERT_NO_MSG(srv->stop != NULL);
+		srv->stop(srv, onoff_stop_notify);
+	} else if (notify) {
+		notify_one(srv, cli, 0);
+	}
+
+	return rv;
+}
+
+static void onoff_reset_notify(struct onoff_service *srv,
+			       int res)
+{
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+	sys_slist_t clients = srv->clients;
+
+	/* If the reset failed clear the transition flag but otherwise
+	 * leave the state unchanged.
+	 *
+	 * If it was successful clear the reference count and all
+	 * flags except capability flags.
+	 */
+	if (res < 0) {
+		srv->flags &= ~SERVICE_STATE_TRANSITION;
+	} else {
+		__ASSERT_NO_MSG(srv->refs == 0U);
+		srv->refs = 0U;
+		srv->flags &= SERVICE_CONFIG_FLAGS;
+	}
+
+	sys_slist_init(&srv->clients);
+
+	k_spin_unlock(&srv->lock, key);
+
+	notify_all(srv, &clients, res);
+}
+
+int z_impl_onoff_service_reset(struct onoff_service *srv,
+			       struct onoff_client *cli)
+{
+	if (srv->reset == NULL) {
+		return -ENOTSUP;
+	}
+
+	bool no_wait = false;
+	bool reset = false;
+	int rv = validate_args(srv, cli, &no_wait);
+
+	if (rv != 0) {
+		return rv;
+	}
+
+	/* Reject if in isr but reset not isr-callable */
+	if (k_is_in_isr()
+	    && ((srv->flags & ONOFF_SERVICE_ISR_CALLABLE_RESET)
+		== 0)) {
+		return -EWOULDBLOCK;
+	}
+
+	/* Reject if in no-wait condition and reset could
+	 * suspend.
+	 */
+	if (no_wait
+	    && ((srv->flags & ONOFF_SERVICE_SUSPENDABLE_RESET)
+		!= 0)) {
+		return -EWOULDBLOCK;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+
+	if ((srv->flags & ONOFF_SERVICE_HAS_ERROR) == 0) {
+		rv = -EINVAL;
+		goto out;
+	}
+
+	if ((srv->flags & SERVICE_STATE_TRANSITION) == 0) {
+		reset = true;
+		srv->flags |= SERVICE_STATE_TRANSITION;
+	}
+	sys_slist_append(&srv->clients, &cli->node);
```

Comment:
This was answered above: the client list is managed atomically so that when new clients come in they're only appended if they're asking for the currently pending operation, and all clients are notified as soon as the operation completes, and after the list has been reset for the next operation.